File: 02_hot100/twoSum2.cpp
// https://leetcode-cn.com/problems/two-sum
/**
 * 哈希表法，时间复杂度O(N)，空间复杂度O(N)
 */
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        vector<int> res;
        unordered_map<int, int> hashtable;
        for (int i = 0; i < nums.size(); i++) {
            auto got = hashtable.find(target-nums[i]);
            if(got != hashtable.end()) {
                res.push_back(got->second);
                res.push_back(i);
            }
            // hashtable.insert(make_pair(nums[i], i));
            hashtable[nums[i]] = i;
        }
        return res;
    }
};  

File: 02_hot100/to be renewal/twoSum2.java

// https://leetcode-cn.com/problems/two-sum
import java.util.HashMap;
import java.util.Map;
/**
 * 哈希表法，时间复杂度O(N)，空间复杂度O(N)
 */
class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> hashtable = new HashMap<Integer, Integer>();
        for (int i = 0; i < nums.length; i++) {
            if (hashtable.containsKey(target - nums[i])) {
                return new int[] {hashtable.get(target - nums[i]), i};
            }
            // 注意这里存放的键值对顺序
            hashtable.put(nums[i], i);
        }
        return new int[0];
    }
}